src/lib.rs

```diff
@@ -1,55 +1,97 @@
-//! Implements the Aleph BFT Consensus protocol as a "finality gadget". The [Consensus] struct
-//! requires access to an [Environment] object which black-boxes the network layer and gives
-//! appropriate access to the set of available blocks that we need to make consensus on.
+//! Implements the Aleph BFT Consensus protocol as a "finality gadget". The [Member] struct
+//! requires access to a network layer, a cryptographic primitive, and a data provided that
+//! gives appropriate access to the set of available data that we need to make consensus on.
 
 use codec::{Decode, Encode};
-use futures::{Future, Sink, Stream};
-use log::{debug, error};
+use futures::Future;
 use std::{
     fmt::{Debug, Display},
-    hash::Hash,
-};
-use tokio::{
-    sync::{mpsc, oneshot},
-    time::Duration,
+    hash::Hash as StdHash,
 };
+use tokio::sync::mpsc;
 
-use crate::{creator::Creator, extender::Extender, syncer::Syncer, terminal::Terminal};
+use crate::nodes::{NodeCount, NodeIndex, NodeMap};
 
-pub use crate::nodes::{NodeCount, NodeIndex, NodeMap};
+pub use member::{Config, Member};
 
+mod bft;
+mod consensus;
 mod creator;
 mod extender;
+mod member;
 pub mod nodes;
 mod syncer;
 mod terminal;
 mod testing;
+mod units;
+
+pub trait DataProvider<Data> {
+    type Error: Debug;
+    fn get_data(&self) -> Data;
+    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error>;
+}
+
+pub trait KeyBox<Signature>: MyIndex {
+    fn sign(&self, msg: &Vec<u8>) -> Signature;
+    fn verify(&self, msg: &Vec<u8>, sgn: &Signature) -> bool;
+}
```

Comment:
Not sure what this trait represents, but this verify() seems strange in this place? What's its purpose?

Reply:
I believe this represents a keychain (all public keys and your private key), so `verify` actually verifies that a signature is correct.

Reply:
Ok, now I see, I have written some comments below about that -- I would add add another argument to verify() being the NodeIndex of the expected signer. Otherwise this forces the Signature to carry the id. In case you want Signature to carry the id, then there should be something like RawSignature that does not carry the id, and we should then always put a RawSignature in structs that are serialized and sent over network.